src/QsCompiler/SyntaxProcessor/CapabilityInference/FeatureAnalyzer.fs

```diff
@@ -0,0 +1,94 @@
+
+type Feature =
+    | Fail
+    | IndeterminateLoop
+    | MultipleReturns
+    | NontrivialCallee
+    | DefaultArray
+
+let createPattern range feature =
+    let capability = TargetCapability.withClassical ClassicalCapability.full TargetCapability.bottom
+
+    let diagnose (target: Target) =
+        let range = QsNullable.defaultValue Range.Zero range
+
+        if TargetCapability.subsumes target.Capability capability then
+            None
+        else
+            let description =
+                match feature with
+                | Fail -> "fail statement"
+                | IndeterminateLoop -> "repeat or while loop"
+                | MultipleReturns -> "multiple returns"
+                | NontrivialCallee -> "callee that is not a global identifier or functor"
+                | DefaultArray -> "default-initialized array constructor"
+
+            let args = [ target.Name; description ]
+            QsCompilerDiagnostic.Error(ErrorCode.UnsupportedClassicalCapability, args) range |> Some
+
+    {
+        Capability = capability
+        Diagnose = diagnose
+        Properties = ()
+    }
+
+let analyzer (action: SyntaxTreeTransformation -> _) : _ seq =
+    let transformation = LocatingTransformation TransformationOptions.NoRebuild
+    let patterns = ResizeArray()
+    let mutable numReturns = 0
+
+    transformation.Statements <-
+        { new StatementTransformation(transformation, TransformationOptions.NoRebuild) with
+            override _.OnStatement statement =
+                transformation.OnRelativeLocation statement.Location |> ignore
+                let range = (transformation.Offset, statement.Location) ||> QsNullable.Map2(fun o l -> o + l.Range)
+
+                match statement.Statement with
+                | QsFailStatement _ -> createPattern range Fail |> patterns.Add
+                | QsRepeatStatement _
+                | QsWhileStatement _ -> createPattern range IndeterminateLoop |> patterns.Add
+                | QsReturnStatement _ ->
+                    numReturns <- numReturns + 1
+                    if numReturns > 1 then createPattern range MultipleReturns |> patterns.Add
+                | _ -> ()
+
+                base.OnStatement statement
+        }
+
+    transformation.Expressions <-
+        { new ExpressionTransformation(transformation, TransformationOptions.NoRebuild) with
+            override _.OnTypedExpression expression =
+                let range = QsNullable.Map2(+) transformation.Offset expression.Range
+
+                match expression.Expression with
+                | CallLikeExpression (callee, _) ->
+                    let isTrivial =
+                        function
+                        | InvalidExpr
+                        | Identifier (GlobalCallable _, _)
+                        | Identifier (InvalidIdentifier, _)
+                        | AdjointApplication _
+                        | ControlledApplication _ -> true
```

Comment:
This is now actually too permissive; You would need to look into whether the inner item is a local callable or not.

Reply:
This function does not need to recurse on the expression, since it is used on the next line with `callee.Exists(isTrivial >> not)`, which does the recursion implicitly. There is an existing test case that checks this:

```qsharp
operation CallFunctorOfExpression() : Unit {
    use q = Qubit();
    Adjoint [Unitary][0](q);
}
```

but I'll add your test case as well.